Reflection_Engine/Query/BaseTypes.cs

```diff
@@ -0,0 +1,68 @@
+/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2022, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Base.Attributes;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+
+namespace BH.Engine.Reflection
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        private static Dictionary<Type, List<Type>> m_cachedBaseTypes = new Dictionary<Type, List<Type>>();
+
+        [Description("Gets all implemented interfaces and the base type (if present) of a given type.")]
+        [Input("type", "Type whose base types will be returned.")]
+        [Output("baseTypes", "All implemented interfaces and the base type (if present) of the input type.")]
+        public static List<Type> BaseTypes(this Type type)
+        {
+            if (type == null)
+                return new List<Type>();
+
+            List<Type> baseTypes;
+
+            if (m_cachedBaseTypes.TryGetValue(type, out baseTypes))
+                return baseTypes;
+            else
+                baseTypes = new List<Type>();
+
+            baseTypes.AddRange(type.GetInterfaces());
+
+            Type baseType = type.BaseType;
+            if (baseType != null)
+                baseTypes.Add(baseType);
+
+            m_cachedBaseTypes[type] = baseTypes;
+
+            return baseTypes;
+        }
+    }
```

Comment:
this might be fixed by above comment, but needs ending `/***************************************************/` after the last method